src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/SelectExpandNode.cs

```diff
@@ -84,24 +82,102 @@ public SelectExpandNode(IEdmStructuredType structuredType, ODataSerializerContex
+        /// Gets the list of EDM navigation properties to be expand referenced in the response.
+        /// keeping this is only for non-breaking changes, This should be replaced by "ReferencedProperties" later.
         /// </summary>
-        /// <param name="selectExpandClause">The parsed $select and $expand query options.</param>
-        /// <param name="structuredType">The structural type of the resource that would be written.</param>
-        /// <param name="model">The <see cref="IEdmModel"/> that contains the given structural type.</param>
-        /// <param name="expandedReference">a boolean value indicating whether it's expanded reference.</param>
-        internal SelectExpandNode(SelectExpandClause selectExpandClause, IEdmStructuredType structuredType, IEdmModel model, bool expandedReference)
-            : this()
+        public ISet<IEdmNavigationProperty> ReferencedNavigationProperties
         {
-            Initialize(selectExpandClause, structuredType, model, false);
+            get
+            {
+                if (ReferencedProperties == null)
+                {
+                    return null;
+                }
+
+                return new HashSet<IEdmNavigationProperty>(ReferencedProperties.Keys);
+            }
+        }
+
+        /// <summary>
+        /// Gets the list of EDM nested properties (complex or collection of complex) to be included in the response.
+        /// keeping this is only for non-breaking changes, This should be replaced by "SelectedComplexes".
+        /// </summary>
+        public ISet<IEdmStructuralProperty> SelectedComplexProperties
+        {
+            get
+            {
+                if (SelectedComplexes == null)
+                {
+                    return null;
+                }
+
+                return new HashSet<IEdmStructuralProperty>(SelectedComplexes.Keys);
+            }
         }
 
+        /// <summary>
+        /// Gets the list of EDM structural properties (primitive, enum or collection of them) to be included in the response.
+        /// It could be null if there's no property selected.
+        /// </summary>
+        public ISet<IEdmStructuralProperty> SelectedStructuralProperties { get; internal set; }
+
+        /// <summary>
+        /// Gets the list of EDM navigation properties to be included as links in the response. It could be null.
+        /// </summary>
+        public ISet<IEdmNavigationProperty> SelectedNavigationProperties { get; internal set; }
+
+        /// <summary>
+        /// Gets the list of Edm structural properties (complex or complex collection) to be included in the response.
+        /// The key is the Edm structural property.
+        /// The value is the potential sub select item.
+        /// </summary>
+        public IDictionary<IEdmStructuralProperty, PathSelectItem> SelectedComplexes { get; internal set; }
```

Comment:
>SelectedComplexes [](start = 67, length = 17)

Don't love the name. Wish we didn't already use SelectedComplexProperties. We should create a V8 work item to remove SelectedComplexProperties and rename this property. In the meantime, maybe call it something like SelectedStructuredProperties?

Reply:
There's a public property named "SelectedStructuredProperties" existed. 

I'd to keep this name unchanged, as just like you mentioned, i created an issue to track the renaming in the v8 at: [https://github.com/OData/WebApi/issues/1976](https://github.com/OData/WebApi/issues/1976)




---
In reply to: [355020521](https://github.com/OData/WebApi/pull/1941#discussion_r355020521) [](ancestors = 355020521)

Reply:
Proposal: rename the new property to SelectedComplexTypedProperties and mark SelectedComplexProperties as obsolete (referencing this new property in the comment)

---
In reply to: [355613891](https://github.com/OData/WebApi/pull/1941#discussion_r355613891) [](ancestors = 355613891,355020521)

Reply:
Done

---
In reply to: [356250680](https://github.com/OData/WebApi/pull/1941#discussion_r356250680) [](ancestors = 356250680,355613891,355020521)